### PR TITLE
Fix serialization and add new impls

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

Fixes some incorrect serialization/deserialization, and adds impls for a few more structs.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Review by ValarDragon — APPROVED** [review]
LGTM! Thanks for fixing the bugs, and adding BTreeSet

Left a general comment regarding a question for serialization guarantees, which we can move to a separate issue

File: serialize/src/lib.rs [hunk 1]
```diff
@@ -559,11 +624,166 @@ where
...
+                K::deserialize_unchecked(&mut reader)?,
+                V::deserialize_unchecked(&mut reader)?,
+            );
+        }
+        Ok(map)
+    }
+}
+
+impl<T: CanonicalSerialize> CanonicalSerialize for BTreeSet<T> {
+    fn serialize<W: Write>(&self, mut writer: W) -> Result<(), SerializationError> {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply serialize/src/lib.rs:654]: err this should have been a comment on the line below, that serialized btree size as a u64 😓

[the remark this answers is not in the document]
Comment (Pratyush) [reply serialize/src/lib.rs:654]: I think the primary motivation is simplicity and correctness. Performance overhead of variable-size decoding would be dwarfed by cryptography-related costs.

[the remark this answers is not in the document]
Comment (ValarDragon) [reply serialize/src/lib.rs:654]: That makes sense, I'll do a pass on the serialization lib to add more docs to it

**Pull request merged**